app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,48 @@
+class PasswordResetsController < ApplicationController
+  require_unauthenticated_access
+  before_action :ensure_manual_login_allowed
+
+  layout "public"
+
+  def new
+  end
+
+  def create
+    os = request.operating_system
+    browser = request.browser
+    SendResetPasswordEmailJob.perform_later(params[:email], os, browser) # queue as a job to avoid timing attacks
+
+    redirect_to '/login', notice: 'If an account with that email was found, we have sent a link to reset the password'
+  end
+
+  def edit
+    @person = find_signed_person(params[:token])
+  end
+
+  def update
+    @person = find_signed_person(params[:token])
+    password_credential = @person&.personable&.password_credential
+
+    if password_credential&.update(password_params)
+      redirect_to '/login', notice: 'Your password was reset succesfully. Please sign in.'
+    else
+      render 'edit', alert: 'There was an error resetting your password'
+    end
+  end
+
+  private
+
+  def find_signed_person(token)
+    Person.find_signed!(token, purpose: Rails.application.config.password_reset_token_purpose)
+  rescue ActiveSupport::MessageVerifier::InvalidSignature
+    redirect_to '/login', alert: 'Your token has expired. Please try again'
+  end
```

Comment:
yeah it only throws the InvalidSignature exception unless it's for the right record, but the record isn't found (I guess if it's been deleted). From the docs:

> Works like [find_signed](https://api.rubyonrails.org/v7.1.3.4/classes/ActiveRecord/SignedId/ClassMethods.html#method-i-find_signed), but will raise an ActiveSupport::MessageVerifier::InvalidSignature exception if the signed_id has either expired, has a purpose mismatch, is for another record, or has been tampered with. It will also raise an ActiveRecord::RecordNotFound exception if the valid signed id can’t find a record.

from: https://api.rubyonrails.org/v7.1.3.4/classes/ActiveRecord/SignedId/ClassMethods.html#method-i-find_signed-21

Figured I'd just let rails handle the RecordNotFound and return a 404 if the person record is actually missing

Reply:
Oh, got it, yes that seems like a good solution if the NotFound will bubble up and abort controller execution.

Let's be sure and represent that as a test case in the controller tests for this. I like how adding test cases serves as a bit of self-documenting code (e.g. a test communications "we thought about this situation, here is how it works")